src/merkle_tree/mod.rs

```diff
@@ -1,253 +1,403 @@
+                    &left_bytes,
+                    &right_bytes,
+                )?
             }
-            upper_bound = start_index;
         }
-        // Finished computing actual tree.
-        // Now, we compute the dummy nodes until we hit our HEIGHT goal.
-        let mut cur_height = tree_height;
-        let mut padding_tree = Vec::new();
-        let mut cur_hash = tree[0].clone();
-        let root_hash = if cur_height < Self::HEIGHT as usize {
-            while cur_height < (Self::HEIGHT - 1) as usize {
-                cur_hash =
-                    hash_inner_node::<P::H>(&parameters, &cur_hash, &empty_hash, &mut buffer)?;
-                padding_tree.push((cur_hash.clone(), empty_hash.clone()));
-                cur_height += 1;
-            }
-            hash_inner_node::<P::H>(&parameters, &cur_hash, &empty_hash, &mut buffer)?
-        } else {
-            cur_hash
-        };
-        end_timer!(new_time);
 
         Ok(MerkleTree {
-            tree,
-            padding_tree,
-            parameters,
-            root: Some(root_hash),
+            leaf_nodes,
+            non_leaf_nodes,
+            height: tree_height,
+            two_to_one_hash_param: two_to_one_hash_param.clone(),
+            leaf_hash_param: leaf_hash_param.clone(),
         })
     }
 
-    #[inline]
-    pub fn root(&self) -> Digest<P> {
-        self.root.clone().unwrap()
+    /// returns the root fo the merkle tree
+    pub fn root(&self) -> TwoToOneDigest<P> {
+        self.non_leaf_nodes[0].clone()
     }
 
-    pub fn generate_proof<L: ToBytes>(
-        &self,
-        index: usize,
-        leaf: &L,
-    ) -> Result<Path<P>, crate::Error> {
-        let prove_time = start_timer!(|| "MerkleTree::GenProof");
-        let mut path = Vec::new();
-
-        let mut buffer = [0u8; 128];
-        let leaf_hash = hash_leaf::<P::H, _>(&self.parameters, leaf, &mut buffer)?;
-        let tree_height = tree_height(self.tree.len());
-        let tree_index = convert_index_to_last_level(index, tree_height);
-        let empty_hash = hash_empty::<P::H>(&self.parameters)?;
-
-        // Check that the given index corresponds to the correct leaf.
-        if leaf_hash != self.tree[tree_index] {
-            return Err(Error::IncorrectLeafIndex(tree_index).into());
-        }
+    /// returns height of the merkle tree
+    pub fn height(&self) -> usize {
+        self.height
+    }
+
+    /// Returns the authentication path from leaf at `index` to root
+    pub fn generate_proof(&self, index: usize) -> Result<Path<P>, crate::Error> {
+        // gather basic tree information
+        let tree_height = tree_height(self.non_leaf_nodes.len() + self.leaf_nodes.len());
 
-        // Iterate from the leaf up to the root, storing all intermediate hash values.
-        let mut current_node = tree_index;
+        // Get Leaf hash, and leaf sibling hash,
+        let leaf_index_in_tree = convert_index_to_last_level(index, tree_height);
```

Comment:
This would also limit us to have `index` <= `2**16` and `tree_height` <= 16 (roughly)

Reply:
I'm not sure why `index <= 2**16`. `index` in this context is the position of a leaf and we want to map it to position in a tree represented by a level-order list. For example, if `index` is `2**31 - 1`, and height is `32`, then `(1 << (tree_height - 1))` is `2**31`, so the return value is `2**31 + 2**31 - 1` = `2**32 - 1`. 

For convenience, Here's the code for `convert_index_to_last_level`:
```rust
#[inline]
fn convert_index_to_last_level(index: usize, tree_height: usize) -> usize {
    index + (1 << (tree_height - 1)) - 1
}
```

Reply:
Yes, sorry, it's `2**31`. I got confused with numbers. That's ok. If the max size of the tree ever becomes a problem, as a library user, we can find a way around

Reply:
Thanks! I will mark this resolved